lab 2 и lab3  Миронюк Ангелина Алексеевна УВВ-112

AngelinaMironyuk opened the pull request:
Реализовать класс Bankomat, моделирующий работу банкомата. В классе должны содержаться поля для хранения идентификационного номера банкомата, информации о текущей сумме денег, оставшейся в банкомате, минимальной и максимальной суммах, которые позволяется снять клиенту в один день. Сумма денег представляется полями-номиналами 10-1000 (см. задание 13). Реализовать метод инициализации банкомата, метод загрузки купюр в банкомат и метод снятия определенной суммы денег. Метод снятия денег должен выполнять проверку на корректность снимаемой суммы: она не должна быть меньше минимального значения и не должна превышать максимальное значение. Метод toString() должен преобразовать в строку сумму денег, оставшуюся в банкомате.

AngelinaMironyuk committed:
Добавлено API и файл main.cpp

AngelinaMironyuk committed:
Добавлено API и файл Bankomat.cpp

AngelinaMironyuk committed:
Создано API класса  и файл Bankomat.cpp

AngelinaMironyuk committed:
Удалила мусорные файлы

AngelinaMironyuk committed:
Удалены {} #1

AngelinaMironyuk committed:
Подключен Bankomat.h в Bankomat.cpp #3

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+{
+public:
+	Bankomat();
+	Bankomat(const int nominals, const int identificationNumber, const int expense, const int max,  const int min) ;
```
стоит подумать о беззнаковом типе. Кажется, что многие поля для данной предметной области не могут быть меньше нуля

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+
+	/**
+	 * \brief метод, возвращающий минимальную сумму денег позволенную снять клиенту в один день
+	 * \return минимальная сумма денег позволенная снять клиенту в один день
```
все-таки за один раз... иначе, строго `0`

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	 * \brief метод, позволяющий определить номинал купюры 
+	 */
+	void SetNominals(int nominals)
+	{
```
реализация в методов должна быть в файле `Bankomat.cpp`

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	 * \brief метод, позволяющий инициализировать банкомат
+	 * \return »инициализация банкомата
+	 */
+	bool initializationBankomat() const ;
```
не понимаю, как Вы себе этот метод представляете. Кажется, что автор задания имел в виду конструкторы

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	/**
+	 * \brief метод, позволяющий загрузить купюры в банкомат
+	 */
+	Bankomat& loadMoneyInBankomat(const Bankomat& other) const ;
```
1. просто `loadMoney`
2. почему Вы копируете объект, а "не кладете купюры"?
3. кажется, что нужен [`class enum`](https://en.cppreference.com/w/cpp/language/enum) для купюр... или структура
4. Вы точно хотите реализовать [`Fluent Interface`](https://ru.wikipedia.org/wiki/Fluent_interface)?

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	/**
+	 * \brief метод, позволяющий снять деньги
+	 */
+	Bankomat& withdrawMoney(const Bankomat& other) const ;
```
опять же не понятно, как второй объект, поступающий на вход функции, разъяснит моему объекту, сколько денег "отдать".
п.4 - ? точно?

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	/**
+	 * \brief метод, позволяющий привести значение к строковому типу
+	 */
+	toString() const ;
```
метод должен что-то выдавать в качестве результата...

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,78 @@
+	toString() const ;
+
+
+private:
```
Нужно проанализировать предметную область. Четко понять, какими типами удобнее будет пользоваться. Просто целые числа - точно нет!

MarinaPaley reviewed (changes requested):
Сначала думаем над `API`. Запускаем код, чтобы он просто собирался. А только потом реализуем!
Реализацию не смотрела

AngelinaMironyuk committed:
#3 Добавлена реализация

MarinaPaley reviewed (changes requested):
Лучше не стало. Исправьте предыдущие замечания.

AngelinaMironyuk committed:
Исправлены ошибки  #5

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,80 @@
+
+
+private:
+
```
докуметация должна быть для каждого члена класса

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,80 @@
+	unsigned long minCash;
+
+
+	const unsigned short TEN = 10;
```
введите `class enum Bill`. тогда Вы сожете для хранения купюр использовать структуру ствндартной библиотеки STL -> `std::map<Bill, size_t> repository;`

MarinaPaley commented on lab1/Bankomat.h:
```diff
@@ -0,0 +1,80 @@
+	unsigned short oneThousand;
+
+	unsigned long maxCash;
+	unsigned long minCash;
```
`maxCash` - сколько можно снять за один день
`minBill` - минимальная купюра

MarinaPaley commented on lab1/Bankomat.cpp:
```diff
@@ -0,0 +1,205 @@
+	this->maxCash = maxCash;
+	this->minCash = minCash;
+
+	setTens(ten);
```
проще сделать "хранилище" через `map<Key, Value>`

MarinaPaley commented on lab1/Bankomat.cpp:
```diff
@@ -0,0 +1,205 @@
+	return sumInCash;
+}
+
+std::string Bankomat::withdrawMoney(unsigned long cash)
```
подумать еще... страшная реализия

MarinaPaley commented on lab1/Bankomat.cpp:
```diff
@@ -0,0 +1,205 @@
+	in >> oneHundred;
+	in >> fiveHundred;
+	in >> oneThousand;
+
```
вызываем конструктор

MarinaPaley reviewed (changes requested):
Переименуйте в `ATM`
Примерная реалиция (сложнее) https://replit.com/@ConstPhi/ATM#main.cpp

AngelinaMironyuk committed:
Исправлены ошибки #5

1.Класс переименован в ATM
2. Добавлен enum для хранения наминалов банкнот
3. Скорректированы конструкторы
4. Оптимизирован алгоритм разбиения суммы на банкноты

AngelinaMironyuk committed:
Создана API и реализация для класса Triad #6

AngelinaMironyuk committed:
Улушения #5
